pkg/providers/instance/aksmachineinstanceutils.go

```diff
@@ -17,12 +17,274 @@ limitations under the License.
+	nodePoolName, ok := aksMachine.Properties.Tags[launchtemplate.NodePoolTagKey]
+	if ok && *nodePoolName != "" {
+		nodePool := &karpv1.NodePool{}
+		if err := kubeClient.Get(ctx, types.NamespacedName{Name: *nodePoolName}, nodePool); err != nil {
+			return nil, err
+		}
+		return nodePool, nil
+	}
+
+	return nil, errors.NewNotFound(schema.GroupResource{Group: coreapis.Group, Resource: "nodepools"}, "")
+}
+
+// ASSUMPTION: NodeClaim name is in the format of <NodePool name>-<hash suffix>
+// If total length exceeds AKS machine name limit, the exceeded part will be replaced with another deterministic hash.
+// E.g., "thisisalongnodepoolname-a1b2c" --> "thisisalongnoz9y8x7-a1b2c"
+func GetAKSMachineNameFromNodeClaimName(nodeClaimName string) (string, error) {
+	const maxAKSMachineNameLength = 35 // Defined by AKS machine API.
+	const prefixHashLength = 6         // The length of the hashed part replacing the exceeded part of the prefix.
+	// If 6, given alphanumeric hash, there will be a total of 36^6 = 2,176,782,336 combinations.
+
+	if len(nodeClaimName) <= maxAKSMachineNameLength {
+		// Safe to use the whole name
+		return nodeClaimName, nil
+	}
+	splitted := strings.Split(nodeClaimName, "-")
+	// Combine the parts except the last one (NodeClaim hash suffix)
+	prefix := strings.Join(splitted[:len(splitted)-1], "-")
+	suffix := "-" + splitted[len(splitted)-1]
+
+	// Keep the legit part of the prefix intact, but hash the rest
+	// ASSUMPTION: prefix length is at least 6 characters at this point (which means suffix length is not too large)
+	// At the time of writing, suffix length is 6 (e.g., "-a1b2c"). This is unlikely to change.
+	keepPrefixLength := maxAKSMachineNameLength - len(suffix) - prefixHashLength
+	prefixToKeep := prefix[:keepPrefixLength]
+	prefixToHash := prefix[keepPrefixLength:]
+	hashedPrefixToHash, err := utils.GetAlphanumericHash(prefixToHash, 6)
+	if err != nil {
+		return "", fmt.Errorf("failed to hash exceeded AKS machine name prefix %q: %w", prefixToHash, err)
+	}
+
+	hashTrimmedPrefix := prefixToKeep + hashedPrefixToHash
+	return hashTrimmedPrefix + suffix, nil
+}
+
+// GetAKSMachineNameFromNodeClaim extracts the AKS machine name from the NodeClaim annotations
+// Returns false if the annotation is not present or the value is empty, which can indicate that the NodeClaim is not associated with an AKS machine.
+func GetAKSMachineNameFromNodeClaim(nodeClaim *karpv1.NodeClaim) (string, bool) {
+	// ASSUMPTION: A NodeClaim is associated with an AKS machine iff the annotation is present (e.g., annotated when creating NodeClaim from a AKS machine instance).
+	// ASSUMPTION: If exists, the annotation is always set to the correct AKS machine resource ID.
+	if aksMachineResourceID, ok := nodeClaim.Annotations[v1beta1.AnnotationAKSMachineResourceID]; ok {
+		splitted := strings.Split(aksMachineResourceID, "/")
+		aksMachineName := splitted[len(splitted)-1] // The last part of the resource ID is the AKS machine name
```

Comment:
Let's maybe use "split" for variable name? Actually, would this work instead (and avoid allocation)?

```suggestion
        aksMachineName := aksMachineResourceID[strings.LastIndex(aksMachineResourceID, "/")+1:]
```